services/worker-manager/src/providers/azure.js

```diff
@@ -451,49 +671,158 @@ class AzureProvider extends Provider {
+        workerPoolId,
+      }, true);
+
+      if (!workerPool) {
+        return; // In this case, the workertype has been deleted so we can just move on
+      }
+
+      if (this.errors[workerPoolId].length) {
+        await Promise.all(this.errors[workerPoolId].map(error => workerPool.reportError(error)));
+      }
+    }));
   }
 
-  async _removeWorker({resourceGroupName, location, vm, ip, nic, disk}) {
-    // clean up related resources, continue on errors
-    // delete VM, IP, NIC, and disk
-    const ignoreNotFound = async (promise) => {
+  /*
+   * removeResource attempts to delete a resource and verify deletion
+   */
+  async removeResource({client, worker, resourceType}) {
+    if (!_.has(worker.providerData, resourceType)) {
+      throw new Error(`Error removing worker: providerData does not contain resourceType ${resourceType}`);
+    }
+    let typeData = worker.providerData[resourceType];
+
+    let shouldDelete = false;
+    // lookup resource by name
+    if (!typeData.id) {
       try {
-        await promise;
+        let {provisioningState} = await this._enqueue('query', () => client.get(
+          worker.providerData.resourceGroupName,
+          typeData.name,
+        ));
+        // resource could be successful, failed, etc.
+        // we have not yet tried to delete the resource
+        if (!(['Deleting', 'Deallocating', 'Deallocated'].includes(provisioningState))) {
+          shouldDelete = true;
+        }
       } catch (err) {
-        if (err.code === 404) {
-          return; // Nothing to do, it is already gone
+        if (err.statusCode === 404) {
+          // if we check for `true` we repeat lots of GET requests
+          // resource has been deleted and isn't in the API or never existed
+          await worker.modify(w => {
+            w.providerData[resourceType].id = false;
+          });
+          return true;
         }
         throw err;
       }
-    };
-    // VM, disk, and NIC _should_ be able to be deleted in parallel
-    // if we get "in use" failures for NIC/disk they will be retried
-    await Promise.all([
-      ignoreNotFound(this._enqueue('query', () => this.computeClient.virtualMachines.deleteMethod(
-        resourceGroupName,
-        vm.name,
-      ))),
-      ignoreNotFound(this._enqueue('query', () => this.networkClient.networkInterfaces.deleteMethod(
-        resourceGroupName,
-        nic.name,
-      ))),
-      await ignoreNotFound(this._enqueue('query', () => this.computeClient.disks.deleteMethod(
-        resourceGroupName,
-        disk.name,
-      ))),
-    ]);
-    // the public IP cannot be deleted as long as it is attached
-    // can become orphaned if we try to delete before the NIC is deleted
-    await ignoreNotFound(this._enqueue('query', () => this.networkClient.publicIPAddresses.deleteMethod(
-      resourceGroupName,
-      ip.name,
-    )));
+    }
+
+    // NB: possible resource leak if we don't require `return true`
```

Comment:
..`return true` where?  I don't see it below here..

Reply:
From this function, i.e. what I mean by that is that the caller should check the return value to make sure we return true, otherwise the deletion might not have finished. Just awaiting the function might have resources in a still deleting state, even though the `id` has been set to `false`.

For example, we check `vmDeleted`:

```
let vmDeleted = await this.removeResource({
  worker,
  client: this.computeClient.virtualMachines,
  resourceType: 'vm',
});
if (!vmDeleted || worker.providerData.vm.id) {
  return state;
}
```